pygatt/bled112_backend.py

```diff
@@ -0,0 +1,1977 @@
+        # registered here will go undetected.
+        self._lib.ble_rsp_attclient_attribute_write +=\
+            self._ble_rsp_attclient_attribute_write
+        self._lib.ble_rsp_attclient_find_information +=\
+            self._ble_rsp_attclient_find_information
+        self._lib.ble_rsp_attclient_read_by_handle +=\
+            self._ble_rsp_attclient_read_by_handle
+        self._lib.ble_rsp_connection_disconnect +=\
+            self._ble_rsp_connection_disconnect
+        self._lib.ble_rsp_connection_get_rssi +=\
+            self._ble_rsp_connection_get_rssi
+        self._lib.ble_rsp_gap_connect_direct +=\
+            self._ble_rsp_gap_connect_direct
+        self._lib.ble_rsp_gap_discover +=\
+            self._ble_rsp_gap_discover
+        self._lib.ble_rsp_gap_end_procedure +=\
+            self._ble_rsp_gap_end_procedure
+        self._lib.ble_rsp_gap_set_mode +=\
+            self._ble_rsp_gap_set_mode
+        self._lib.ble_rsp_gap_set_scan_parameters +=\
+            self._ble_rsp_gap_set_scan_parameters
+        self._lib.ble_rsp_sm_delete_bonding +=\
+            self._ble_rsp_sm_delete_bonding
+        self._lib.ble_rsp_sm_encrypt_start +=\
+            self._ble_rsp_sm_encrypt_start
+        self._lib.ble_rsp_sm_get_bonds +=\
+            self._ble_rsp_sm_get_bonds
+        self._lib.ble_rsp_sm_set_bondable_mode +=\
+            self._ble_rsp_sm_set_bondable_mode
+
+        # Register event handlers. Note: the packets of any event not registered
+        # here will go undetected.
+        self._lib.ble_evt_attclient_attribute_value +=\
+            self._ble_evt_attclient_attribute_value
+        self._lib.ble_evt_attclient_find_information_found +=\
+            self._ble_evt_attclient_find_information_found
+        self._lib.ble_evt_attclient_procedure_completed +=\
+            self._ble_evt_attclient_procedure_completed
+        self._lib.ble_evt_connection_status +=\
+            self._ble_evt_connection_status
+        self._lib.ble_evt_connection_disconnected +=\
+            self._ble_evt_connection_disconnected
+        self._lib.ble_evt_gap_scan_response +=\
+            self._ble_evt_gap_scan_response
+        self._lib.ble_evt_sm_bond_status +=\
+            self._ble_evt_sm_bond_status
+        self._lib.ble_evt_sm_bonding_fail +=\
+            self._ble_evt_sm_bonding_fail
+
+        # Start logging
+        self._loglock.acquire()
+        self._logger.info("BLED112Backend on %s", serial_port)
+        self._loglock.release()
+
+        # Run the receiver thread
+        if run:
+            self.run()
+
+    def bond(self):
+        """
+        Create a bond and encrypted connection with the device.
+
+        This requires that a connection is already extablished with the device.
+        """
+        # Get locks
+        self._main_thread_cond.acquire()
+        self._loglock.acquire()
+
+        # Make sure there is a connection
```

Comment:
Get locks, drop locks make private functions. updated in branch add-switchable-backends_reviewed.